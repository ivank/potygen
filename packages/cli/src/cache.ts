import { existsSync, readFileSync, writeFileSync, statSync, mkdirSync } from 'fs';
import { dirname } from 'path';
import { createHash } from 'crypto';

export class CachedFileParser<TResult> {
  constructor(public store = new Map<string, TResult>(), public process: (content: string, path: string) => TResult) {}

  public get(path: string): TResult {
    const content = readFileSync(path, 'utf-8');
    const key = createHash('md5').update(content).digest('hex');
    const cachedResult = this.store.get(key);
    if (cachedResult) {
      return cachedResult;
    } else {
      const result = this.process(content, path);
      this.store.set(key, result);
      return result;
    }
  }
}

export class CachedProcessParser<TResult> {
  constructor(public store = new Map<string, TResult>(), public process: (content: string, path: string) => TResult) {}

  public get(path: string): TResult {
    const content = readFileSync(path, 'utf-8');
    const key = createHash('md5').update(content).digest('hex');
    const cachedResult = this.store.get(key);
    if (cachedResult) {
      return cachedResult;
    } else {
      const result = this.process(content, path);
      this.store.set(key, result);
      return result;
    }
  }
}

export class CacheStore {
  public store = new Map<string, number>();

  constructor(public fileName: string, public enabled: boolean = false, public cacheClear: boolean = false) {
    if (this.enabled && !this.cacheClear) {
      if (existsSync(fileName)) {
        this.store = new Map(Object.entries(JSON.parse(readFileSync(fileName, 'utf-8'))));
      }
    }
  }

  public shouldParseFile(path: string) {
    if (!this.enabled) {
      return true;
    } else {
      const mtime = statSync(path).mtime.getTime();
      const currentMtime = this.store.get(path);
      if (currentMtime && mtime === currentMtime) {
        return false;
      }
    }
    return true;
  }

  public cacheFile(path: string) {
    const mtime = statSync(path).mtime.getTime();
    this.store.set(path, mtime);
  }

  public save() {
    if (this.enabled) {
      mkdirSync(dirname(this.fileName), { recursive: true });
      writeFileSync(this.fileName, JSON.stringify(Object.fromEntries(this.store)), 'utf-8');
    }
  }
}
