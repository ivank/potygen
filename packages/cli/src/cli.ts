import { Client } from 'pg';
import { pipeline } from 'stream';
import { promisify } from 'util';
import { SqlRead, QueryLoader } from './traverse';

const asyncPipeline = promisify(pipeline);

const main = async () => {
  const db = new Client({
    database: 'schemes',
    user: 'schemes',
    password: 'dev-pass',
    host: '127.0.0.1',
    port: 5432,
    ssl: false,
  });
  try {
    await db.connect();

    const sqls = new SqlRead('src/**/*.query.ts', '../../../schemes/packages/api');
    const sink = new QueryLoader(
      db,
      '../../../schemes/packages/api',
      '{{root}}/src/queries/__psql-ts__/{{name}}.queries.ts',
    );

    await asyncPipeline(sqls, sink);
  } catch (error) {
    console.log(String(error));
  } finally {
    await db.end();
  }
};

main();
