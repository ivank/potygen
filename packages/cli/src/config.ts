import { Record, Optional, Static, String, Boolean } from 'runtypes';

export const Config = Record({
  files: Optional(String),
  watch: Optional(Boolean),
  root: Optional(String),
  connection: Optional(String),
  template: Optional(String),
  verbose: Optional(Boolean),
  silent: Optional(Boolean),
  typePrefix: Optional(String),
  preload: Optional(Boolean),
  cache: Optional(Boolean),
  cacheFile: Optional(String),
  cacheClear: Optional(Boolean),
});

export type ConfigType = Static<typeof Config>;
export type FullConfigType = Required<ConfigType>;

export const toConfig = (config: unknown): FullConfigType => ({
  files: '**/*.sql',
  cacheFile: '.cache/potygen.cache',
  root: process.cwd(),
  preload: false,
  template: '{{dir}}/{{name}}.queries.ts',
  connection: 'postgres://localhost:5432/db',
  watch: false,
  verbose: false,
  silent: false,
  cache: false,
  cacheClear: false,
  typePrefix: '',
  ...Config.check(config),
});
