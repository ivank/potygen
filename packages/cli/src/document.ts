import { document, printDocument, DocumentContext, Type, withIdentifier } from '@ovotech/ts-compose';
import { TypeNode } from 'typescript';
import { isLoadedArrayType, isLoadedUnionType, isLoadedLiteralType, isLoadedValuesPick } from './guards';
import { LoadedQuery, LoadedValuesPick, LoadedType } from './types';

const toPropertyType = (type: LoadedType | LoadedValuesPick): TypeNode => {
  if (isLoadedValuesPick(type)) {
    return Type.Array(
      Type.TypeLiteral({
        props: type.items.map((item) => Type.Prop({ name: item.name, type: toPropertyType(item.value) })),
      }),
    );
  } else if (isLoadedArrayType(type)) {
    return Type.Array(toPropertyType(type.items));
  } else if (isLoadedUnionType(type)) {
    return Type.Union(type.items.map(toPropertyType));
  } else if (isLoadedLiteralType(type)) {
    return Type.Literal(type.value);
  } else {
    switch (type.type) {
      case 'Date':
        return Type.Referance('Date');
      case 'boolean':
        return Type.Boolean;
      case 'json':
        return Type.Unknown;
      case 'null':
        return Type.Null;
      case 'number':
        return Type.Number;
      case 'string':
        return Type.String;
      case 'unknown':
        return Type.Unknown;
    }
  }
};

export const toQueryTypescriptTypes = (context: DocumentContext, query: LoadedQuery) => {
  const paramsContext = withIdentifier(
    context,
    Type.Interface({
      name: 'QueryParams',
      isExport: true,
      props: query.params.map(({ name, type }) =>
        Type.Prop({ name, type: toPropertyType(type), isOptional: type.optional }),
      ),
    }),
  );

  const resultContext = withIdentifier(
    paramsContext,
    Type.Interface({
      name: 'QueryResult',
      isExport: true,
      props: query.result.map(({ name, type }) =>
        Type.Prop({ name, type: toPropertyType(type), isOptional: type.optional }),
      ),
    }),
  );

  return document(
    resultContext,
    Type.Interface({
      name: 'Query',
      isExport: true,
      props: [
        Type.Prop({ name: 'params', type: Type.Referance('QueryParams') }),
        Type.Prop({ name: 'result', type: Type.Referance('QueryResult') }),
      ],
    }),
  );
};

export const toQueryTypescript = (loadedQuery: LoadedQuery): string => {
  return printDocument(toQueryTypescriptTypes({}, loadedQuery));
};
