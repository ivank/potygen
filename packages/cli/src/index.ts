export {
  DataTable,
  DataFunction,
  DataEnum,
  DataView,
  DataComposite,
  LoadedDataTable,
  LoadedDataComposite,
  DataViewRaw,
  DataViewParsed,
  LoadedDataView,
  LoadedDataFunction,
  LoadedDataEnum,
  Data,
  LoadedData,
  LoadedParam,
  LoadedResult,
  LoadedQueryInterface,
  LoadedFunction,
  LoadedComposite,
  LoadedSourceTable,
  LoadedSourceQuery,
  LoadedSourceView,
  LoadedSourceValues,
  LoadedSource,
  LoadedContext,
  TemplateTagQuery,
  ParsedTypescriptFile,
  ParsedSqlFile,
  ParsedFile,
  LoadedTemplateTagQuery,
  LoadedTypescriptFile,
  LoadedSqlFile,
  LoadedFile,
  Logger,
  LoadContext,
  QualifiedName,
} from './types';

export {
  isDataTable,
  isDataFunction,
  isDataEnum,
  isLoadedDataTable,
  isLoadedDataFunction,
  isLoadedDataEnum,
  isLoadedDataView,
  isLoadedDataComposite,
} from './guards';

export { toLoadedContext, loadQueryInterfacesData, toLoadedQueryInterface, loadData, extractDataSources } from './load';

export { Config, ConfigType, FullConfigType, toConfig } from './config';

export { emitLoadedFile, toTypeSource } from './emit';
export { glob } from './glob';
