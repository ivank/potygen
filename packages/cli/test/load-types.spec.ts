import { parser } from '@psql-ts/ast';
import { Client } from 'pg';
import { loadQueryInterface, loadQueryInterfaces } from '../src/load';
import { toQueryInterface } from '@psql-ts/query';

let db: Client;

describe('Query Interface', () => {
  beforeAll(async () => {
    db = new Client({ database: 'sql-ast', user: 'sql-ast', password: 'dev-pass' });
    await db.connect();
  });

  afterAll(async () => {
    await db.end();
  });

  it.each<[string, string]>([
    ['function result single', `SELECT ABS(integer_col) FROM all_types`],
    ['function result double', `SELECT ABS(ABS(integer_col)) FROM all_types`],
    ['nested function guess type', `SELECT ABS(ARRAY_LENGTH(ARRAY_AGG(integer_col), 1)) FROM all_types GROUP BY id`],
    [
      'nested function explicit type',
      `SELECT ARRAY_LENGTH(ARRAY_AGG(integer_col), 1)::int[] FROM all_types GROUP BY id`,
    ],
    ['operators integer', `SELECT integer_col + integer_col AS "test1" FROM all_types WHERE id = $id`],
    ['operators string', `SELECT character_col + integer_col AS "test1" FROM all_types WHERE character_col = $text`],
    ['different result types', `SELECT * FROM all_types`],
    ['enum', `SELECT 'Pending'::account_levelisation_state`],
    ['enum column', `SELECT state FROM account_levelisations`],
    ['simple', `SELECT id, character_col FROM all_types WHERE id = :id`],
    ['coalesce 1', `SELECT COALESCE(id, character_col) FROM all_types`],
    ['coalesce 2', `SELECT COALESCE(character_col, id) FROM all_types`],
    ['coalesce 3', `SELECT COALESCE(12, character_col, id) FROM all_types`],
    ['coalesce 4', `SELECT COALESCE(integer_col, character_col) FROM all_types`],
    ['parameter coalesce', `SELECT character_col FROM all_types WHERE integer_col > COALESCE($id, 2)`],
  ])('Should convert %s sql (%s)', async (_, sql) => {
    const ast = parser(sql);
    const query = toQueryInterface(ast!);
    const loadedQuery = await loadQueryInterface(db, query);

    expect(loadedQuery.queryInterface).toMatchSnapshot();
  });

  it('Should load multple queries', async () => {
    const queries = [
      `SELECT ABS(integer_col) FROM all_types`,
      `SELECT ABS(ABS(integer_col)) FROM all_types`,
      `SELECT ABS(ARRAY_LENGTH(ARRAY_AGG(integer_col), 1)) FROM all_types GROUP BY id`,
      `SELECT ARRAY_LENGTH(ARRAY_AGG(integer_col), 1)::int[] FROM all_types GROUP BY id`,
      `SELECT integer_col + integer_col AS "test1" FROM all_types WHERE id = $id`,
      `SELECT character_col + integer_col AS "test1" FROM all_types WHERE character_col = $text`,
      `SELECT * FROM all_types`,
      `SELECT 'Pending'::account_levelisation_state`,
      `SELECT state FROM account_levelisations`,
      `SELECT id, character_col FROM all_types WHERE id = :id`,
    ].map((sql) => {
      const ast = parser(sql);
      return toQueryInterface(ast!);
    });

    const loadedQueries = await loadQueryInterfaces(db, queries);
    const individuallyLoaded = (await Promise.all(queries.map((query) => loadQueryInterface(db, query)))).map(
      ({ queryInterface }) => queryInterface,
    );

    expect(loadedQueries.queryInterfaces).toEqual(individuallyLoaded);
  });
});
