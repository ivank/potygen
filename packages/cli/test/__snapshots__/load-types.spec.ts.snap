// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query Interface Should convert coalesce 1 sql (SELECT COALESCE(id, character_col) FROM all_types) 1`] = `
Object {
  "params": Array [],
  "results": Array [
    Object {
      "name": "coalesce",
      "type": Object {
        "items": Array [
          Object {
            "type": "Number",
          },
          Object {
            "nullable": true,
            "type": "String",
          },
        ],
        "nullable": false,
        "type": "UnionConstant",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert coalesce 2 sql (SELECT COALESCE(character_col, id) FROM all_types) 1`] = `
Object {
  "params": Array [],
  "results": Array [
    Object {
      "name": "coalesce",
      "type": Object {
        "items": Array [
          Object {
            "nullable": true,
            "type": "String",
          },
          Object {
            "type": "Number",
          },
        ],
        "nullable": false,
        "type": "UnionConstant",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert coalesce 3 sql (SELECT COALESCE(12, character_col, id) FROM all_types) 1`] = `
Object {
  "params": Array [],
  "results": Array [
    Object {
      "name": "coalesce",
      "type": Object {
        "items": Array [
          Object {
            "literal": 12,
            "type": "Number",
          },
          Object {
            "nullable": true,
            "type": "String",
          },
          Object {
            "type": "Number",
          },
        ],
        "nullable": false,
        "type": "UnionConstant",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert coalesce 4 sql (SELECT COALESCE(integer_col, character_col) FROM all_types) 1`] = `
Object {
  "params": Array [],
  "results": Array [
    Object {
      "name": "coalesce",
      "type": Object {
        "items": Array [
          Object {
            "nullable": true,
            "type": "Number",
          },
          Object {
            "nullable": true,
            "type": "String",
          },
        ],
        "nullable": true,
        "type": "UnionConstant",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert composite type from table sql (SELECT (item).supplier_id FROM all_types) 1`] = `
Object {
  "params": Array [],
  "results": Array [
    Object {
      "name": "supplier_id",
      "type": Object {
        "nullable": true,
        "type": "Number",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert composite type from view sql (SELECT (item).supplier_id FROM all_types_view) 1`] = `
Object {
  "params": Array [],
  "results": Array [
    Object {
      "name": "supplier_id",
      "type": Object {
        "nullable": true,
        "type": "Number",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert different result types sql (SELECT * FROM all_types) 1`] = `
Object {
  "params": Array [],
  "results": Array [
    Object {
      "name": "id",
      "type": Object {
        "type": "Number",
      },
    },
    Object {
      "name": "not_null",
      "type": Object {
        "type": "Number",
      },
    },
    Object {
      "name": "default_not_null",
      "type": Object {
        "type": "Number",
      },
    },
    Object {
      "name": "bigint_col",
      "type": Object {
        "nullable": true,
        "type": "BigInt",
      },
    },
    Object {
      "name": "bigserial_col",
      "type": Object {
        "type": "BigInt",
      },
    },
    Object {
      "name": "bit_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "bit_varying_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "boolean_col",
      "type": Object {
        "nullable": true,
        "type": "Boolean",
      },
    },
    Object {
      "name": "box_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "bytea_col",
      "type": Object {
        "type": "Buffer",
      },
    },
    Object {
      "name": "character_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "character_varying_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "cidr_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "circle_col",
      "type": Object {
        "items": Array [
          Object {
            "name": "x",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "y",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "radius",
            "type": Object {
              "type": "Number",
            },
          },
        ],
        "type": "ObjectLiteralConstant",
      },
    },
    Object {
      "name": "date_col",
      "type": Object {
        "nullable": true,
        "type": "Date",
      },
    },
    Object {
      "name": "double_col",
      "type": Object {
        "nullable": true,
        "type": "Number",
      },
    },
    Object {
      "name": "inet_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "integer_col",
      "type": Object {
        "nullable": true,
        "type": "Number",
      },
    },
    Object {
      "name": "interval_col",
      "type": Object {
        "items": Array [
          Object {
            "name": "years",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "months",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "days",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "hours",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "minutes",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "seconds",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "milliseconds",
            "type": Object {
              "type": "Number",
            },
          },
        ],
        "type": "ObjectLiteralConstant",
      },
    },
    Object {
      "name": "json_col",
      "type": Object {
        "nullable": true,
        "type": "Json",
      },
    },
    Object {
      "name": "jsonb_col",
      "type": Object {
        "nullable": true,
        "type": "Json",
      },
    },
    Object {
      "name": "line_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "lseg_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "macaddr_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "money_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "numeric_col",
      "type": Object {
        "nullable": true,
        "type": "Number",
      },
    },
    Object {
      "name": "path_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "pg_lsn_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "point_col",
      "type": Object {
        "items": Array [
          Object {
            "name": "x",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "y",
            "type": Object {
              "type": "Number",
            },
          },
        ],
        "type": "ObjectLiteralConstant",
      },
    },
    Object {
      "name": "polygon_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "real_col",
      "type": Object {
        "nullable": true,
        "type": "Number",
      },
    },
    Object {
      "name": "smallint_col",
      "type": Object {
        "nullable": true,
        "type": "Number",
      },
    },
    Object {
      "name": "smallserial_col",
      "type": Object {
        "type": "Number",
      },
    },
    Object {
      "name": "serial_col",
      "type": Object {
        "type": "Number",
      },
    },
    Object {
      "name": "text_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "time_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "time_with_time_zone_col",
      "type": Object {
        "nullable": true,
        "type": "Date",
      },
    },
    Object {
      "name": "timestamp_col",
      "type": Object {
        "nullable": true,
        "type": "Date",
      },
    },
    Object {
      "name": "timestamp_with_time_zone_col",
      "type": Object {
        "nullable": true,
        "type": "Date",
      },
    },
    Object {
      "name": "tsquery_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "tsvector_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "txid_snapshot_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "uuid_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "xml_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "item",
      "type": Object {
        "attributes": Object {
          "name": Object {
            "nullable": true,
            "type": "String",
          },
          "price": Object {
            "nullable": true,
            "type": "Number",
          },
          "supplier_id": Object {
            "nullable": true,
            "type": "Number",
          },
        },
        "name": "inventory_item",
        "schema": "public",
        "type": "CompositeConstant",
      },
    },
    Object {
      "name": "state",
      "type": Object {
        "items": Array [
          Object {
            "literal": "Active",
            "type": "String",
          },
          Object {
            "literal": "Pending",
            "type": "String",
          },
          Object {
            "literal": "Dispute",
            "type": "String",
          },
          Object {
            "literal": "Closed",
            "type": "String",
          },
        ],
        "nullable": true,
        "type": "UnionConstant",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert enum column sql (SELECT state FROM account_levelisations) 1`] = `
Object {
  "params": Array [],
  "results": Array [
    Object {
      "name": "state",
      "type": Object {
        "items": Array [
          Object {
            "literal": "Pending",
            "type": "String",
          },
          Object {
            "literal": "Done",
            "type": "String",
          },
        ],
        "nullable": false,
        "type": "UnionConstant",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert enum sql (SELECT 'Pending'::account_levelisation_state) 1`] = `
Object {
  "params": Array [],
  "results": Array [
    Object {
      "name": "row",
      "type": Object {
        "items": Array [
          Object {
            "literal": "Pending",
            "type": "String",
          },
          Object {
            "literal": "Done",
            "type": "String",
          },
        ],
        "type": "UnionConstant",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert function result double sql (SELECT ABS(ABS(integer_col)) FROM all_types) 1`] = `
Object {
  "params": Array [],
  "results": Array [
    Object {
      "name": "abs",
      "type": Object {
        "type": "Number",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert function result single sql (SELECT ABS(integer_col) FROM all_types) 1`] = `
Object {
  "params": Array [],
  "results": Array [
    Object {
      "name": "abs",
      "type": Object {
        "type": "Number",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert nested function explicit type sql (SELECT ARRAY_LENGTH(ARRAY_AGG(integer_col), 1)::int[] FROM all_types GROUP BY id) 1`] = `
Object {
  "params": Array [],
  "results": Array [
    Object {
      "name": "array",
      "type": Object {
        "items": Object {
          "type": "Number",
        },
        "type": "ArrayConstant",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert nested function guess type sql (SELECT ABS(ARRAY_LENGTH(ARRAY_AGG(integer_col), 1)) FROM all_types GROUP BY id) 1`] = `
Object {
  "params": Array [],
  "results": Array [
    Object {
      "name": "abs",
      "type": Object {
        "type": "Number",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert operators integer sql (SELECT integer_col + integer_col AS "test1" FROM all_types WHERE id = $id) 1`] = `
Object {
  "params": Array [
    Object {
      "name": "id",
      "type": Object {
        "nullable": true,
        "type": "OptionalConstant",
        "value": Object {
          "type": "Number",
        },
      },
    },
  ],
  "results": Array [
    Object {
      "name": "test1",
      "type": Object {
        "nullable": true,
        "type": "Number",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert operators string sql (SELECT character_col + integer_col AS "test1" FROM all_types WHERE character_col = $text) 1`] = `
Object {
  "params": Array [
    Object {
      "name": "text",
      "type": Object {
        "nullable": true,
        "type": "OptionalConstant",
        "value": Object {
          "nullable": true,
          "type": "String",
        },
      },
    },
  ],
  "results": Array [
    Object {
      "name": "test1",
      "type": Object {
        "nullable": true,
        "type": "Number",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert parameter coalesce sql (SELECT character_col FROM all_types WHERE integer_col > COALESCE($id, 2)) 1`] = `
Object {
  "params": Array [
    Object {
      "name": "id",
      "type": Object {
        "nullable": true,
        "type": "OptionalConstant",
        "value": Object {
          "nullable": true,
          "type": "Number",
        },
      },
    },
  ],
  "results": Array [
    Object {
      "name": "character_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert parameter required sql (SELECT character_col FROM all_types WHERE integer_col > $id!) 1`] = `
Object {
  "params": Array [
    Object {
      "name": "id",
      "type": Object {
        "nullable": false,
        "type": "OptionalConstant",
        "value": Object {
          "nullable": true,
          "type": "Number",
        },
      },
    },
  ],
  "results": Array [
    Object {
      "name": "character_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert select where sql (SELECT character_col FROM all_types WHERE integer_col > COALESCE($id, 2)) 1`] = `
Object {
  "params": Array [
    Object {
      "name": "id",
      "type": Object {
        "nullable": true,
        "type": "OptionalConstant",
        "value": Object {
          "nullable": true,
          "type": "Number",
        },
      },
    },
  ],
  "results": Array [
    Object {
      "name": "character_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert simple sql (SELECT id, character_col FROM all_types WHERE id = :id) 1`] = `
Object {
  "params": Array [
    Object {
      "name": "id",
      "type": Object {
        "nullable": true,
        "type": "OptionalConstant",
        "value": Object {
          "type": "Number",
        },
      },
    },
  ],
  "results": Array [
    Object {
      "name": "id",
      "type": Object {
        "type": "Number",
      },
    },
    Object {
      "name": "character_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert sum sql (SELECT SUM(integer_col) FROM all_types) 1`] = `
Object {
  "params": Array [],
  "results": Array [
    Object {
      "name": "sum",
      "type": Object {
        "type": "BigInt",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert views types sql (SELECT * FROM all_types_view) 1`] = `
Object {
  "params": Array [],
  "results": Array [
    Object {
      "name": "id",
      "type": Object {
        "type": "Number",
      },
    },
    Object {
      "name": "not_null",
      "type": Object {
        "type": "Number",
      },
    },
    Object {
      "name": "default_not_null",
      "type": Object {
        "type": "Number",
      },
    },
    Object {
      "name": "bigint_col",
      "type": Object {
        "nullable": true,
        "type": "BigInt",
      },
    },
    Object {
      "name": "bigserial_col",
      "type": Object {
        "type": "BigInt",
      },
    },
    Object {
      "name": "bit_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "bit_varying_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "boolean_col",
      "type": Object {
        "nullable": true,
        "type": "Boolean",
      },
    },
    Object {
      "name": "box_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "bytea_col",
      "type": Object {
        "type": "Buffer",
      },
    },
    Object {
      "name": "character_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "character_varying_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "cidr_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "circle_col",
      "type": Object {
        "items": Array [
          Object {
            "name": "x",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "y",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "radius",
            "type": Object {
              "type": "Number",
            },
          },
        ],
        "type": "ObjectLiteralConstant",
      },
    },
    Object {
      "name": "date_col",
      "type": Object {
        "nullable": true,
        "type": "Date",
      },
    },
    Object {
      "name": "double_col",
      "type": Object {
        "nullable": true,
        "type": "Number",
      },
    },
    Object {
      "name": "inet_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "integer_col",
      "type": Object {
        "nullable": true,
        "type": "Number",
      },
    },
    Object {
      "name": "interval_col",
      "type": Object {
        "items": Array [
          Object {
            "name": "years",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "months",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "days",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "hours",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "minutes",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "seconds",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "milliseconds",
            "type": Object {
              "type": "Number",
            },
          },
        ],
        "type": "ObjectLiteralConstant",
      },
    },
    Object {
      "name": "json_col",
      "type": Object {
        "nullable": true,
        "type": "Json",
      },
    },
    Object {
      "name": "jsonb_col",
      "type": Object {
        "nullable": true,
        "type": "Json",
      },
    },
    Object {
      "name": "line_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "lseg_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "macaddr_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "money_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "numeric_col",
      "type": Object {
        "nullable": true,
        "type": "Number",
      },
    },
    Object {
      "name": "path_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "pg_lsn_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "point_col",
      "type": Object {
        "items": Array [
          Object {
            "name": "x",
            "type": Object {
              "type": "Number",
            },
          },
          Object {
            "name": "y",
            "type": Object {
              "type": "Number",
            },
          },
        ],
        "type": "ObjectLiteralConstant",
      },
    },
    Object {
      "name": "polygon_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "real_col",
      "type": Object {
        "nullable": true,
        "type": "Number",
      },
    },
    Object {
      "name": "smallint_col",
      "type": Object {
        "nullable": true,
        "type": "Number",
      },
    },
    Object {
      "name": "smallserial_col",
      "type": Object {
        "type": "Number",
      },
    },
    Object {
      "name": "serial_col",
      "type": Object {
        "type": "Number",
      },
    },
    Object {
      "name": "text_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "time_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "time_with_time_zone_col",
      "type": Object {
        "nullable": true,
        "type": "Date",
      },
    },
    Object {
      "name": "timestamp_col",
      "type": Object {
        "nullable": true,
        "type": "Date",
      },
    },
    Object {
      "name": "timestamp_with_time_zone_col",
      "type": Object {
        "nullable": true,
        "type": "Date",
      },
    },
    Object {
      "name": "tsquery_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "tsvector_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "txid_snapshot_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "uuid_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "xml_col",
      "type": Object {
        "nullable": true,
        "type": "String",
      },
    },
    Object {
      "name": "item",
      "type": Object {
        "attributes": Object {
          "name": Object {
            "nullable": true,
            "type": "String",
          },
          "price": Object {
            "nullable": true,
            "type": "Number",
          },
          "supplier_id": Object {
            "nullable": true,
            "type": "Number",
          },
        },
        "name": "inventory_item",
        "schema": "public",
        "type": "CompositeConstant",
      },
    },
    Object {
      "name": "state",
      "type": Object {
        "items": Array [
          Object {
            "literal": "Active",
            "type": "String",
          },
          Object {
            "literal": "Pending",
            "type": "String",
          },
          Object {
            "literal": "Dispute",
            "type": "String",
          },
          Object {
            "literal": "Closed",
            "type": "String",
          },
        ],
        "nullable": true,
        "type": "UnionConstant",
      },
    },
  ],
}
`;
